### PR TITLE
Return exit code from docker exec

### LINKED ISSUE #250: No error return if Command::Exec execution fails in dockerman
To reproduce: take the following snapshot of gumpi: https://github.com/golemfactory/gumpi/tree/4537ac1aa3cb10de6e77fe7ac2a685fda891a2a2

Modify `examples/heat_solver.toml`: `s/--initial/foobar/`

Execute: (assuming that the hub is running on 127.0.0.1, with 6 cores available on the machine)
```
RUST_LOG=gumpi=debug,gu_client=debug cargo run -- --hub 127.0.0.1:61622 --job examples/heat_solver.toml -n 6 --noclean
```
The `error!` statement should be executed: https://github.com/golemfactory/gumpi/blob/4537ac1aa3cb10de6e77fe7ac2a685fda891a2a2/src/session/mpi.rs#L264
but in fact the `and_then` branch is taken: https://github.com/golemfactory/gumpi/blob/4537ac1aa3cb10de6e77fe7ac2a685fda891a2a2/src/session/mpi.rs#L280

This is due to the fact that `do_exec` in dockerman apparently doesn't handle the return exit code:
https://github.com/golemfactory/golem-unlimited/blob/release/0.2/gu-provider/src/dockerman.rs#L84-L115

### PR DESCRIPTION
**Pull request by filipgolem**
Closes https://github.com/golemfactory/golem-unlimited/issues/250.

### CODE REVIEW

File: gu-provider/src/dockerman.rs [hunk 1]
```diff
@@ -106,16 +106,31 @@ impl DockerSession {
...
+            .and_then(|(stream, id)| {
+                stream
+                    .fold(String::new(), |mut s, (_t, it)| {
+                        match std::str::from_utf8(it.into_bytes().as_ref()) {
+                            Ok(chunk_str) => s.push_str(chunk_str),
+                            Err(_) => (),
+                        };
+                        Ok::<String, String>(s)
+                    })
+                    .map_err(|e| format!("{}", e))
```
Comment (marmistrz) [inline gu-provider/src/dockerman.rs:124]: Could we move this `map_err` before the final `and_then` and get rid of the `map_err` for `check_exit_status`?

Comment (marmistrz) [reply gu-provider/src/dockerman.rs:124]: You can also check if `.map_err(ToString::to_string)` works as expected. According to the [doc](https://doc.rust-lang.org/std/string/trait.ToString.html) it should, but I'm not sure if it worked as expected for me.

File: gu-provider/src/dockerman.rs [hunk 2]
```diff
@@ -106,16 +106,31 @@ impl DockerSession {
...
+                    .and_then(move |output| {
+                        let exit_status = container_copy
+                            .check_exec_status(&id)
+                            .map_err(|e| format!("{}", e));
+                        exit_status.join(future::ok(output))
+                    })
+                    .and_then(|(status, output)| match status {
+                        0 => future::ok(output),
```
Comment (marmistrz) [inline gu-provider/src/dockerman.rs]: `and_then` expects an `IntoFuture`, so you can use `Ok`/`Err` to be more concise.

File: gu-provider/src/dockerman.rs [hunk 3]
```diff
@@ -106,16 +106,31 @@ impl DockerSession {
...
+                        exit_code => future::err(format!("{}\n{}", output, exit_code)),
```
Comment (marmistrz) [inline gu-provider/src/dockerman.rs]:
```suggestion
                        exit_code => future::err(format!("{}\nExit code: {}", output, exit_code)),
```
We can't really do much better than this until #240 is addressed, but we should be a little more informative to the end-user.

File: gu-provider/src/dockerman.rs [hunk 4]
```diff
@@ -106,16 +106,31 @@ impl DockerSession {
...
+                            Err(_) => (),
+                        };
+                        Ok::<String, String>(s)
+                    })
+                    .map_err(|e| format!("{}", e))
+                    .and_then(move |output| {
+                        let exit_status = container_copy
+                            .check_exec_status(&id)
+                            .map_err(|e| format!("{}", e));
+                        exit_status.join(future::ok(output))
```
Comment (marmistrz) [inline gu-provider/src/dockerman.rs]: You don't need the new variable here, just add the `join` as the tail call.

**Review by marmistrz — CHANGES_REQUESTED** [review]

File: gu-provider/src/dockerman.rs (hunk 4 above)
Comment (filipgolem) [reply gu-provider/src/dockerman.rs]: Fixed

File: gu-provider/src/dockerman.rs (hunk 2 above)
Comment (filipgolem) [reply gu-provider/src/dockerman.rs]: Fixed

File: gu-provider/src/dockerman.rs (hunk 1 above)
Comment (filipgolem) [reply gu-provider/src/dockerman.rs:124]: Fixed

**Pull request merged**